app/controllers/sessions_controller.rb

```diff
@@ -1,7 +1,9 @@
 class SessionsController < ApplicationController
   def new
-    if request.original_url != login_for_cart_url
-      session[:return_to] ||= request.referer
+    if request.path != login_for_cart_path
+      session[:return_to] = request.referer
+    else
+      session[:return_to] = new_order_path
     end
```

Comment:
If a user is not logged in, they visit the login_for_cart_path which goes to the session#new action.

If they click Login from the nav bar they go to the login_path which also goes to the session#new controller. 

the request.path shows what path they took to get to the session#new which we can use to determine how to redirect after they log in